module-1/lab-errhand_listcomp/your-code/main.py

```diff
@@ -254,14 +254,22 @@ def square_it():
 # 22. Find all of the numbers from 1-1000 that are divisible by any single digit besides 1 (2-9). 
```

Comment:
7. It's easier and more readable based on vowels. The example shown in class on Friday
```Python
teststring = 'Find all of the words in a string that are monosyllabic'
vowels = ['a','e','i','o','u',' ']
non_vowels = [letter for letter in teststring if letter.lower() not in vowels]
print(non_vowels)
```
8. It's easier and more readable using isupper() function 
```Python
results = [letter for letter in teststring if letter.isupper()]
print(capital_letters)
print(results)
```
11. Why did you use tolist() function? the output of a list comprehension is already a list. Using numpy is not necessary. Keep things as simple as possible. An example: 
```Python
random_lists = [random.sample(range(101),10) for _ in range (4)]
print(random_lists)
```
15. Notation for exceptions is except followed by the type of error. Single except throws the general type of error. The print statement usually includes the message you want to prompt to you, other analysts, users, .....
```Python 
x = 5
y = 0

try:
    z = x/y
except ZeroDivisionError:4
    print ("Can't divide by Zero!")
finally:
    print ('All Done!')
```
17. You were expected to use try/except blocks to solve this question. An example:
```Python
try:
    x = int( input( "enter a number: " ) )
    y = int( input( "enter another number: " ) )
    print( x, '/', y, '=', x/y )
except ZeroDivisionError:
    print( "Can't divide by 0!" )
except ValueError:
    print( "That doesn't look like a number!" )
except:
    print( "something unexpected happend!" )
```
18. You were expected to use IOError or similar instead of the general Exception Python class
19. You were expected to use more than one exception type. You can use just one try/except block in this case instead of two.  An example: 
```Python
try:
    fp = open('myfile.txt')
    line = f.readline()
    i = int(s.strip())
except (IOError,ValueError) as e:
    print ("Please check the file,either the file is read-only or the data can't be converted to an integer.",e.errno)
except: 
    print ("Unexpected error")
```
20. You were expected to use an AssertionError instead of the general Exception one. 
22. This is the long way, shorter ways could be as follows: 
```Python
results = [number for number in range(1,1001) if True in [True for divisor in range(2,10) if number % divisor == 0]]
print(results)
#Ivan's approach 
results=[[['' if f%e!=0 else e for e in range(2,10)],f] for f in range(2,1001)]
print(results)
#AlbertoG's approach
results = set([x for x in range(1, 1001) for i in range(2, 10) if x%i == 0])
print(results)
```
23. You were expected to customize the error building with Python Class as it was described in the hint. You only need to raise one exception, total marks scored doesn't need an exception in this case. A way could be:
```Python
class UnAcceptedValueError(Exception):   
    def __init__(self, data):    
        self.data = data
    def __str__(self):
        return repr(self.data)

Total_Marks = int(input("Enter Total Marks Scored: "))
try:
    Num_of_Sections = int(input("Enter Num of Sections: "))
    if(Num_of_Sections < 2):
        raise UnAcceptedValueError("Number of Sections can't be less than 2")
except UnAcceptedValueError as e:
    print ("Received error:", e.data)
```